src/did-manager/cheqd-did-provider.ts

```diff
@@ -104,16 +109,336 @@ export type TImportableEd25519Key = Required<Pick<IKey, 'publicKeyHex' | 'privat
 	type: 'Ed25519';
 };
 
+
+export class CheqdProviderError extends Error {
+	constructor(message: string, public readonly errorCode: string) {
+		super(message);
+		this.errorCode = errorCode;
+	}
+}
+
+export const CheqdProviderErrorCodes = {
+	DeactivatedController: 'DeactivatedController',
+	UnresolvedDID: 'UnresolvedDID',
+	EmptyVerificationMethod: 'EmptyVerificationMethod'
+} as const;
+
+export type TPublicKeyEd25519 = Required<Pick<IKey, 'publicKeyHex'>> & {
+	kid: TImportableEd25519Key['publicKeyHex'];
+	type: 'Ed25519';
+};
+
 declare const TImportableEd25519Key: {
 	isTImportableEd25519Key(object: object[]): object is TImportableEd25519Key[];
 };
 
 export type TSupportedKeyType = 'Ed25519' | 'Secp256k1';
 
+export interface ICheqdKey extends IKey {
+	controller?: string
+}
```

Comment:
Let's switch this to a more representative naming:

```suggestion
export interface IKeyWithController extends IKey {
	controller?: string
}
```

This can be defined at the global scope as well, if necessary.
Interfaces are enriched by union of fields defined, which is one of their unique TS feature set.